shopfloor_reception/services/reception.py

```diff
@@ -0,0 +1,951 @@
+        picking = self.env["stock.picking"].browse(picking_id)
+        if all(line.qty_done == 0 for line in picking.move_line_ids):
+            # If no line has been processed, refuse to set the picking as done
+            return self._response_for_select_line(
+                picking, message=self.msg_store.transfer_no_qty_done()
+            )
+        if not confirmation:
+            to_backorder = picking._check_backorder()
+            if to_backorder:
+                # Not all lines are fully done, ask the user to confirm the
+                # backorder creation
+                return self._response_for_select_line(
+                    picking, message=self.msg_store.transfer_confirm_done()
+                )
+            # all lines are done, ask the user to confirm anyway
+            return self._response_for_select_line(
+                picking, message=self.msg_store.need_confirmation()
+            )
+        self._handle_backorder(picking)
+        return self._response_for_select_document(
+            message=self.msg_store.transfer_done_success(picking)
+        )
+
+    def _handle_backorder(self, picking):
+        """This method handles backorders that could be created at picking confirm."""
+        backorders_before = picking.backorder_ids
+        picking._action_done()
+        backorders_after = picking.backorder_ids - backorders_before
+        # Remove user_id on the backorder, if any
+        backorders_after.user_id = False
+
+    def set_lot(
+        self, picking_id, selected_line_ids, lot_name=None, expiration_date=None
+    ):
+        """Set lot and its expiration date
+
+        Input:
+            barcode: The barcode of a lot
+            expiration_date: The expiration_date
+
+        transitions:
+          - select_line: User clicked on back
+          - set_lot: Barcode not found. Ask user to create one from barcode
+          - set_lot: expiration_date has been set on the selected line
+          - set_lot: lot_it has been set on the selected line
+          - set_lot: Error: expiration_date is required
+          - set_quantity: User clicked on the confirm button
+        """
+        picking = self.env["stock.picking"].browse(picking_id)
+        selected_lines = self.env["stock.move.line"].browse(selected_line_ids)
+        message = self._check_picking_status(picking)
+        if message:
+            return self._response_for_set_lot(picking, selected_lines, message=message)
+        if not selected_lines.exists():
+            message = self.msg_store.record_not_found()
+            return self._response_for_set_lot(picking, selected_lines, message=message)
+        lot_model = self.env["stock.production.lot"]
+        search = self._actions_for("search")
+        if lot_name:
+            for line in selected_lines:
+                lot = search.lot_from_scan(
+                    lot_name, products=selected_lines.mapped("product_id")
```

Comment:
For each line you search for all the products of all lines.
I would collect all the line w/out a lot, collect all values for missing lots and create them all at once.
You can map line ids and lots in a dict.

Reply:
We only deal with 1 line at a time now so this is no longer the case.